allennlp/models/model.py

```diff
@@ -309,8 +310,36 @@ def _load(
         # If vocab and model embeddings are in sync, following would be just a no-op.
         model.extend_embedder_vocab()
 
+        # Load state dict. We pass `strict=False` so PyTorch doesn't raise a RuntimeError
+        # if the state dict is missing keys because we handle this case below.
         model_state = torch.load(weights_file, map_location=util.device_mapping(cuda_device))
-        model.load_state_dict(model_state)
+        missing_keys, unexpected_keys = model.load_state_dict(model_state, strict=False)
+
+        # Modules might define a class variable called `authorized_missing_keys`,
+        # a list of regex patterns, that tells us to ignore missing keys that match
+        # any of the patterns.
+        # We sometimes need this in order to load older models with newer versions of AllenNLP.
+
+        def filter_out_authorized_missing_keys(module, prefix=""):
+            nonlocal missing_keys
+            for pat in getattr(module.__class__, "authorized_missing_keys", None) or []:
+                missing_keys = [
+                    k
+                    for k in missing_keys
+                    if k.startswith(prefix) and re.search(pat[len(prefix) :], k) is None
+                ]
+            for name, child in module._modules.items():
+                if child is not None:
+                    filter_out_authorized_missing_keys(child, prefix + name + ".")
+
+        filter_out_authorized_missing_keys(model)
```

Comment:
Does this really have to be recursive? Couldn't we just say that the regex matches a key of the form `"module.submodule1.submodule2"`, and leave it up to the flexibility of regular expressions?

That's how we use the regexes in other scenarios, like for initializers parameter groups.

Reply:
We need to use recursion to find all of the child modules. Is that what you're asking about?

Reply:
We already have all the `missing_keys` in a list, so we could just match the regexes against those, right?

Reply:
Yes, but we don't have the regex patterns, so we need to recursively go through the child modules to collect those.

Reply:
Ah, makes sense now.

Can you just say `getstate(module.__class__, "authorized_missing_keys", [])` instead of `getattr(module.__class__, "authorized_missing_keys", None) or []`?

Reply:
I tried that, but there are some modules defined in the transformers library that have `authorized_missing_keys` set to `None`.